reviewer/reviewer.go

```diff
@@ -0,0 +1,253 @@
+				selections = append(selections, users[i])
+				break
+			}
+		}
+	}
+	return selections
+}
+
+func shoveIntoMap(u []string, m map[string]struct{}) map[string]struct{} {
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+	return m
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, client *github.Client) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		allUsers = shoveIntoMap(child.RequestedUsers, allUsers)
+
+		if len(child.RequestedTeams) > 0 {
+			randomTeam := child.RequestedTeams[r.Intn(len(child.RequestedTeams))]
+			teamInfo := strings.Split(randomTeam, "/")
+			team, _, err := client.Teams.GetTeamBySlug(ctx, teamInfo[0], teamInfo[1])
+			if err != nil {
+				logger.Debug().Err(err).Msgf("Unable to get member listing for team %s", randomTeam)
+				//return nil, errors.Wrapf(err, "Unable to get information for team %s", randomTeam)
+			} else {
+				teamMembers, err := listAllTeamMembers(ctx, client, team)
+				if err != nil {
+					logger.Debug().Err(err).Msgf("Unable to get member listing for team %s", randomTeam)
+					//return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+				} else {
+					allUsers = shoveIntoMap(teamMembers, allUsers)
+				}
+			}
+		}
+
+		if len(child.RequestedOrganizations) > 0 {
+			randomOrg := child.RequestedOrganizations[r.Intn(len(child.RequestedOrganizations))]
+			orgMembers, err := listAllOrgMembers(ctx, client, randomOrg)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Unable to get member listing for org %s", randomOrg)
+			}
+			allUsers = shoveIntoMap(orgMembers, allUsers)
+		}
+
+		if child.RequestedAdmins {
+			repoAdmins, err := listAllCollaborators(ctx, client, prctx.RepositoryOwner(), prctx.RepositoryName(), "admin")
+			if err != nil {
+				return nil, errors.Wrap(err, "Unable to get admin listing")
+			}
+			allUsers = shoveIntoMap(repoAdmins, allUsers)
+		}
+
+		if child.RequestedWriteCollaborators {
+			repoCollaborators, err := listAllCollaborators(ctx, client, prctx.RepositoryOwner(), prctx.RepositoryName(), "write")
+			if err != nil {
+				return nil, errors.Wrap(err, "Unable to get admin listing")
+			}
+			allUsers = shoveIntoMap(repoCollaborators, allUsers)
+		}
+
+		// Remove author before randomly selecting, since github will fail to assign _anyone_
+		if _, ok := allUsers[prctx.Author()]; ok {
+			delete(allUsers, prctx.Author())
+		}
+
+		// Remove any users who aren't collaborators, since github will fail to assign _anyone_
+		for k := range allUsers {
+			adminContrib, err := prctx.IsCollaborator(prctx.RepositoryOwner(), prctx.RepositoryName(), k, common.GithubAdminPermission)
```

Comment:
This is pretty expensive: we probably want to filter ahead of time using the repo collaborators. Alternatively instead just randomly pick and hope things go okay, and if not then run it again.